doc: Ansible/Helm add api command  Proposal for Operator SDK (#2641)
operator-framework/operator-sdk: doc/proposals/ansible-helm-addapi.md

---
title: Ansible/Helm add api command  Proposal for Operator SDK
authors:
  - "@bharathi-tenneti"
reviewers:
  - "@fabianvf"
  - "@cmacedo"
approvers:
  - "@fabianvf"
  - "@jlanford"
  - "@dmesser"
creation-date: 2020-03-10
last-updated: 2020-03-11
status: implementable
---

# Ansible/Helm add api command  Proposal for Operator SDK

## Release Signoff Checklist

- \[x\] Enhancement is `implementable`

> - \[x\] Design details are appropriately documented from clear requirements

Reviewer: I think it would be just checked if the proposal has detailed information over how to do the implementation.
For example, is possible to use the operator-sdk api command that already exists today? When we call the operator-sdk new <project> where is this implementation? Can we use it in the operator-sdk api? Will be required to implement a specific command for Helm and other for Ansible or not? 

Also, I'd add the info regards how we can check the type of the operator whiout the need to use the flag --type

Author: "Also, I'd add the info regards how we can check the type of the operator whiout the need to use the flag --type"
For above comment, The idea is to modify "add api" command, which does not look for --type flag, rather use [this](https://github.com/operator-framework/operator-sdk/blob/master/cmd/operator-sdk/new/cmd.go#L121) to determine the operator type.

- \[ \] Test plan is defined
- \[ \] Accceptance criteria
- \[ \] User-facing documentation is created

## Summary

> In current scenario, users are not able to add additional APIs for Ansible/Helm based opertators using SDK CLI. This change will make it easier to add additional apis using CLI for Ansible and Helm based operators.

Reviewer 2: typo: apis -> APIs to be consistent.

> 

Reviewer: ```suggestion
Currently, users are not able to add additional APIs for Ansible/Helm based-operators via SDK CLI tool. This proposal is over provide this feature via the tool. 

```

Author: I think I will stick with the original sentencing. It conveys the same.

Reviewer: It is not a change, it is a proposal with a propose of providing a new feature to solve a need.

Reviewer 3: it's a proposal to change the existing behavior (the existing behavior being exiting in failure). Adding features is still making a change, I think this phrasing is fine

## Motivation

> For Ansible/Helm operator SDK, user is not able to add additional apis for already scaffolded operators.

Reviewer 2: typo: apis -> APIs

> For Ansible/Helm operator SDK, user is not able to add additional apis for already scaffolded operators.

Reviewer: Here would be the scenario which justifies the implementation. I mean, it is more related to why users would need to add additional APIs.

## Goals

> The goal of the proposal if implemented, is to enable users add additional APIs once the original Ansible/Helm operator is scaffolded.

Reviewer 2: add a comma (,) after proposal before the if. i.e. "The goal of the proposal, if implemented,..."

> ## Non-Goals

Reviewer: ```suggestion
```
We can just remove those ones which are not useful for us here.

> NA

Reviewer: ```suggestion
```

## Proposal

### User Stories 

#### Story 1 - Ansible Operator-sdk

> Enable Ansible operator-sdk to scaffold additional api, once the oroginal Ansible operator project has been created, using bwloq command.

Reviewer 2: typo: api -> API
typo: oroginal -> original

What's `bwloq`?

`operator-sdk add api --kind <kind> --api-version <group/version>`  

> 

Reviewer: I'd recommend write the story following its concept and common practices. See; https://www.mountaingoatsoftware.com/agile/user-stories

> ##### Tasks included:

Reviewer: The following are the list of the files that need to be scaffold. It is not part of the Story. It is detailed information about how the implementation needs to be done.

> * Update ansible role at ../deplo/role.yaml

Reviewer 2: typo: deplo -> deploy

> * Create directpry for new kind under ../roles/

Reviewer 2: typo: directpry > directory

* CRD and CR files are generated at ../deploy/crds/                                   
* Update watches file at ../watches.yaml
* Update playbook with the generated resources (**TBD**)
* Update Molecule tests under ../molecule/ (**TBD**)

#### Story 2 - Helm Operator-sdk

> Enable Helm operator-sdk to scaffold additional api, once the original Helm operator project has been created, using below command, --helm-chart being optional.

Reviewer 2: typo: api -> API

`operator-sdk add api --kind <kind> --api-version <group/version> --helm-chart=stable/nginx` 

##### Tasks included:

> * Update helm role at ../deplo/role.yaml

Reviewer 2: typo: deplo -> deploy

* CRD and CR files are generated at ../deploy/crds/ 
* Charts directory is generated under ../helm-charts
* Update watches file at ../watches.yaml

### Implementation Details/Notes/Constraints
* Analyze how existing Go operator is scaffolding additonal apis.
* Manually add additonal api for an Ansible operator example as PoC.

> * Modify existin "add api" functionality under /cmd/add/api.go, to include ansible workflow.

Reviewer 2: typo: existin -> existing

* Discuss on whether to generate playbook for ansible in case of additional api.
* Discuss on whether to generate molecule tests as well for additional apis.

> 

Reviewer: IMO here we would need answers, for example, is possible to use the operator-sdk api command that already exists today? When we call the operator-sdk new <project> where is this implementation? Can we use it in the operator-sdk api? Will be required to implement a specific command for Helm and other for Ansible or not? 

Also, I'd add the info regards how we can check the type of the operator without the need to use the flag --type

### Risks and Mitigations
NA

> ### Acceptance Criteria

Reviewer: IMO: Acceptance Criteria are part of the Story
It defined the criteria to accept the story as done.

* Documentation is updated with steps to add additonal apis for both Ansible and Helm.
* Flags options which are related to the API scaffold and available for operator-sdk new project --type   also should be provided in the operator-sdk new api for the types

> * Ansible User should be able to scadffold all resources needed for the additional api with below command

Reviewer 2: typo: scadffold -> scaffold

`operator-sdk add api --kind <kind> --api-version <group/version>` 

> * Helm User should be able to scadffold all resources needed for the additional api with below command,   --helm-chart being optionl

Reviewer 2: typo: scadffold -> scaffold

`operator-sdk add api --kind <kind> --api-version <group/version> --helm-chart=stable/nginx` 

> 

Reviewer 2: Remove all these whitespace lines, they're not needed.

last-updated: 2020-03-12
status: implementable
---

# Ansible/Helm add api command proposal for Operator SDK

## Release Signoff Checklist

- \[x\] Enhancement is `implementable`
- \[ \] Design details are appropriately documented from clear requirements
- \[ \] Test plan is defined
- \[ \] Accceptance criteria
- \[ \] User-facing documentation is created

## Summary

The proposal is to enable Ansible/Helm operator users to create additional APIs, through SDK CLI.

## Motivation

Ansible/Helm operator users are not able to create additonal APIs via CLI, once the original project scaffolds. Today, users have to manually add necessary files to the project scaffold.

## Goals

* Ansible/Helm operator developer can use existing SDK CLI commands to create additonal APIs as needed.
* Ansible/Helm operator developer can find supported documentation for the same.

## Proposal

### User Stories 

#### Story 1 - Ansible operator additional API
As an  Ansible operator developer, I would like to scaffold additional api, once the original Ansible operator project has been created, using following command.
`operator-sdk add api --kind <kind> --api-version <group/version>`

> ##### Implementation Details/Notes/Constraints

Reviewer: this section is child of the proposal and not story.

---
title: Ansible/Helm add api command proposal for Operator SDK
authors:
  - "@bharathi-tenneti"
reviewers:
  - "@fabianvf"
  - "@cmacedo"
approvers:
  - "@fabianvf"
  - "@jlanford"
  - "@dmesser"
creation-date: 2020-03-10
last-updated: 2020-03-15
status: implementable
---

# Ansible/Helm add API command proposal for Operator SDK

## Release Signoff Checklist

- \[x\] Enhancement is `implementable`

> - \[ \] Design details are appropriately documented from clear requirements

Reviewer: ```suggestion
- \[x \] Design details are appropriately documented from clear requirements
```

- \[ \] Test plan is defined
- \[ \] Accceptance criteria
- \[ \] User-facing documentation is created

## Summary

The proposal is to enable Ansible/Helm operator developers to create additional APIs, through SDK CLI.

## Motivation

As of today, SDK CLI can be used only to add additonal APIs for Go based Operators.
Ansible/Helm operator developers are not able to create additonal APIs via CLI, once the original project scaffolds. Today, developers have to manually add necessary files to the project scaffold.

## Goals

* Ansible/Helm operator developer can use existing SDK CLI commands to create additonal APIs as needed.
* Ansible/Helm operator developer should be able to use flags necessary for Ansible/Helm as used in `operator-sdk new`CLI. for adding additional APIs as well.
* Ansible/Helm operator developer can find supported documentation for the same.

## Non-Goals

* Updating Molecule tests for additional APIs created for Anible based operators.[**TBD**]
* Generating/Updating Playbook for additional APIs for Ansible based operators. [**TBD**]

> 

Reviewer: Are you ok with @fabianvf, should we not to do both?

Reviewer 3: I'm fine with this, for developers these parts would be the bits they understand, handling the CRD stuff is the most useful

    `operator-sdk add api --kind <kind> --api-version <group/version> [flags]`

##### Acceptance Criteria
* Helm operator developer should be able to scaffold all resources needed for the additional API with any of below commands, 
      `operator-sdk add api --kind <kind> --api-version <group/version> [flags]`
* Flags options available for `operator-sdk new`, should also be made available for `operator-sdk add api`
```
  --api-version string - CRD APIVersion in the format $GROUP_NAME/$VERSION (e.g app.example.com/v1alpha1)
  --kind string - CRD Kind. (e.g AppService)
  --helm-chart string - Initialize helm operator with existing helm chart (<URL>, <repo>/<name>, or local path)
  --helm-chart-repo string - Chart repository URL for the requested helm chart
  --helm-chart-version string - Specific version of the helm chart (default is latest version)
  ```
* Documentation for [SDK CLI reference][sdkclidoc] is updated with steps to add additonal APIs for Helm based operator.
* Documentation is updated for [operator-sdk add api][addapidoc] for helm.

### Implementation Details/Notes/Constraints

> * `operator-sdk new memcached-operator --api-version=cache.example.com/v1alpha1 --kind=Memcached --type=ansible` scaffolds new ansible based operator for the user with given API. Below logic is being used to determine the type of operator.

Reviewer: ```suggestion
The `operator-sdk new memcached-operator --api-version=cache.example.com/v1alpha1 --kind=Memcached --type=ansible` scaffolds new ansible based operator for the user with given API. Below logic is being used to determine the type of operator.
```

```go
	case projutil.OperatorTypeAnsible:
		if err := doAnsibleScaffold(); err != nil {
			return err
		}
	case projutil.OperatorTypeHelm:
		if err := doHelmScaffold(); err != nil {
			return err
		}
```
and subsequently [`func doAnsibleScaffold()`][doansible] or [`func doHelmScaffold()`][dohelm] is being called to perform the scaffold.

> * Currently, `operator-sdk add api` only allows Go based operators to create futher APIs, after the original project is scaffolded. By posing retriction as shown [here][onlygorestriction].

Reviewer: ```suggestion
Currently, `operator-sdk add api` only allows Go-based operators to create further APIs after the original project is scaffolded. By posing restriction as shown [here][onlygorestriction].
```

```go
// Only Go projects can add apis.
	if err := projutil.CheckGoProjectCmd(cmd); err != nil {
		return err
	}
```

> * The proposal is to enhance [`func apiRun(cmd *cobra.Command, args []string)`][addapifunc] to add APIs for Ansible/Helm operators, by re-using pre-existing functions as shown above to check for `--type`, and perform necessary scaffolds for the new resource.

Reviewer: ```suggestion
This proposal is to enhance [`func apiRun(cmd *cobra.Command, args []string)`][addapifunc] to add APIs for Ansible/Helm operators, by re-using pre-existing functions as shown above to check for `--type`, and perform necessary scaffolds for the new resource.
```

> * To this extent, PoCs have been done for both [Ansible][ansiblepoc] and [Helm][helmpoc] by manually adding necessary files in the scaffold.

Reviewer: ```suggestion
* To this extent, PoCs have been done for both [Ansible][ansiblepoc] and [Helm][helmpoc] by manually adding necessary files in the scaffold. See its project layout: 
```

> 

Reviewer: ```suggestion
**Examples**
```

> * Ansible roles scaffold after adding APIs:

Reviewer: ```suggestion
**NOTE**: To test/check the POCs locally used the makefile targets `make install` and `make uninstall`.

* Ansible roles scaffold after adding APIs:
```

│       │   ├── serviceaccount.yaml
│       │   └── tests
│       │       └── test-connection.yaml
│       └── values.yaml
```
* Along with above changes,`/deploy/role.yaml` will be updated to reflect new apiGroups.
``` yaml
- apiGroups:
  - cache.example.com
  resources:
  - '*'
  verbs:
  - create
  - delete
  - get
  - list
  - patch
  - update
  - watch
- apiGroups:
  - app.example.com
  resources:
  - '*'
  verbs:
  - create
  - delete
  - get
  - list
  - patch
  - update
  - watch
```
* CRD and CR files will be generated at `/deploy/crds`, as shown below.
```
── deploy
│   ├── crds
│   │   ├── cache.example.com_memcacheds_crd.yaml
│   │   ├── cache.example.com_v1alpha1_memcached_cr.yaml
│   │   ├── charts.helm.k8s.io_mongodbs_crd.yaml
│   │   ├── charts.helm.k8s.io_v1alpha1_mongodb_cr.yaml
│   │   ├── example.com_nginxes_crd.yaml
│   │   └── example.com_v1alpha1_nginx_cr.yaml
```
* watches.yaml at `/watches.yaml` gets updated with new API resource as shown below.
```yaml
- version: v1alpha1
  group: cache.example.com
  kind: Memcached
  role: /opt/ansible/roles/memcached

- version: v1alpha1
  group: cache.example.com
  kind: Mykind
  role: /opt/ansible/roles/mykind
```

> 

Reviewer: ```suggestion
```

Reviewer: Remove spaces.

> # Ansible/Helm add api command  Proposal for Operator SDK

Reviewer: @bharathi-tenneti the template that we need to use is: https://github.com/operator-framework/operator-sdk/blob/master/doc/proposals/TEMPLATE.md

Author: I was looking at previous proposal docs, https://github.com/operator-framework/operator-sdk/blob/master/doc/proposals/ansible-operator.md

> Status: implementable
> 

- [Background](#background)
- [Goals](#goals)
- [Commands](#commands)
- [Ansible_Changes] (#ansible changes)
- [Helm_Changes] (#helm changes)

> ## Background

Reviewer: Following my suggestion for : https://github.com/operator-framework/operator-sdk/blob/master/doc/proposals/TEMPLATE.md#user-stories-optional
 
**Story**

- I am as an Ansible operator-sdk developer, I'd like to be able to scaffold new API"s by using the command `operator-sdk new Api`
- I am as a Helm operator-sdk developer, I'd like to be able to scaffold new API"s by using the command `operator-sdk new Api`
- I am as a Helm/Ansible operator-sdk developer, I'd like to not use the flag --type when the project is already scaffold. 

**Acceptance Criteria**

- Documentation updated with these steps
- Flags options which are related to the API scaffold and available for operator-sdk new project --type [options] also should be provided in the operator-sdk new api for the types. 
- All required files for the new Kind be a scaffold for an Ansible based-operator when the command `operator-sdk new --api-version=<group>/<version> --kind=<Kind>` is executed for example.
- All required files for the new Kind be scaffolded for a Helm based-operator when the command `operator-sdk new memcached-operator --api-version=<group>/<version> --kind=<Kind> --helm-chart=stable/memcached` is executed for example.
 
PS.: I would split by type, so for example, Ansible would be story 1 and Helm story 2

In current scenario, users are not able to add additional APIs for Ansible/Helm based opertators using SDK CLI.

## Goals

The goal of the proposal if implemented, is to enable users add additional APIs once the original Ansible/Helm operator is scaffolded.

## Commands
We are updating existing commands to accommodate the ansible/helm operator.  Changes to the `cmd\add\api.go` is needed.

`operator-sdk add api --kind <kind> --api-version <group/version>`     

## Ansible Changes

> 

Reviewer: For each new Kind, I understand that a dir is created so should we not create the new dir with the full structure? 

**Example**

1. We created the project with the Memcached API

── roles
│   └── Memcached
│       ├── README.md
│       ├── defaults
│       │   └── main.yml
│       ├── handlers
│       │   └── main.yml
│       ├── meta
│       │   └── main.yml
│       ├── tasks
│       │   └── main.yml
│       ├── tests
│       │   ├── inventory
│       │   └── test.yml
│       └── vars
│           └── main.yml

2. Then, the new Kind AppServer will be added:

├── roles
│   ├── AppServer
│   │   ├── README.md
│   │   ├── defaults
│   │   │   └── main.yml
│   │   ├── handlers
│   │   │   └── main.yml
│   │   ├── meta
│   │   │   └── main.yml
│   │   ├── tasks
│   │   │   └── main.yml
│   │   ├── tests
│   │   │   ├── inventory
│   │   │   └── test.yml
│   │   └── vars
│   │       └── main.yml
│   └── Memcached
│       ├── README.md
│       ├── defaults
│       │   └── main.yml
│       ├── handlers
│       │   └── main.yml
│       ├── meta
│       │   └── main.yml
│       ├── tasks
│       │   └── main.yml
│       ├── tests
│       │   ├── inventory
│       │   └── test.yml
│       └── vars
│           └── main.yml

Also, I understand that watches need to be updated as:

$ cat watches.yaml 
---
- version: v1alpha1
  group: cache.example.com
  kind: Memcached
  role: Memcached

- version: v1alpha1
  group: cache.example.com
  kind: AppServer
  role: AppServer

Author: Yes, Directory under "roles" is needed, looks like this point got missed in the proposed changes.  watches.yaml changes are noted already in the proposed changes

Author: Regarding the molecule directory, @fabianvf WDYT?

Reviewer: I think we can close this one. I made a comment for him in the no-goals and shows that you addressed all already. 👍

* Udate ansible role at ../deplo/role.yaml
* CRD and CR files are generated at ../deploy/crds/                                   
* Update watches file at ../watches.yaml
* Update playbook with the generated resources(TBD)

> 

Reviewer: Did you make a POC with Memcached Sample by adding manually all that is required to do check if will work well? 
Also, it has an option/flag that will do scaffold with watches playbooks, we need to check that as well.
